File: 백준/Gold/15683. 감시/감시.py
import copy

n,m = map(int,input().split())
board = [list(map(int,input().split())) for _ in range(n)]
cctv=[]
for i in range(n):
    for j in range(m):
        if board[i][j] in [1,2,3,4,5]:
            cctv.append((board[i][j],i,j))

dx = [-1, 0, 1, 0]
dy = [0, 1, 0, -1]
direct = { 1: [[1], [2], [3], [4]] , 2: [[1,3], [2,4]], 3: [[1,2], [2,3], [3,4], [4,1]], 4:[[1,2,3], [2,3,4], [3,4,1], [4,1,2]], 5:[[1,2,3,4]] }
min_val=[2147483647]


def search(arr, dir, x, y):
    for i in dir:
        nx=x
        ny=y
        while 1:
            nx+=dx[i-1]
            ny+=dy[i-1]
            if nx<0 or ny<0 or nx>=n or ny>=m:
                break
            if arr[nx][ny]==6:
                break
            if arr[nx][ny]==0:
                arr[nx][ny]=7


def dfs(depth, board):
    if depth==len(cctv):
        cnt=0
        for i in range(n):
            cnt+=board[i].count(0)
        min_val[0]=min(min_val[0],cnt)
        return

    tmp= copy.deepcopy(board)
    now , x , y =cctv[depth]
    for i in direct[now]:
        search(tmp,i,x,y)
        dfs(depth+1,tmp)
        tmp=copy.deepcopy(board)

dfs(0,board)
print(min_val[0])

File: 백준/Gold/15683. 감시/감시.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Main {
	static int[][] board;
	static List<int[]> cctv=new ArrayList<>();
	static int N;
	static int M;
	static int[] dx= {-1,0,1,0};
	static int[] dy= {0,1,0,-1};
	static int[][][] dir= {
			{},
			{{0},{1},{2},{3}},
			{{0,2},{1,3}},
			{{0,1},{1,2},{2,3},{3,0}},
			{{0,1,3},{0,1,2},{1,2,3},{2,3,0}},
			{{0,1,2,3}}
	};
	static int min=Integer.MAX_VALUE;
	
    public static void main(String[] args) throws IOException {
        BufferedReader br=new BufferedReader(new InputStreamReader(System.in));
        String[] xy=br.readLine().split(" ");
        N=Integer.parseInt(xy[0]);
        M=Integer.parseInt(xy[1]);
        int[][] board = new int[N][M];
        for(int i=0; i<N; i++) {
        	String[] tmp=br.readLine().split(" ");
        	for(int j=0; j<tmp.length; j++) {
        		board[i][j]=Integer.parseInt(tmp[j]);
        		if(board[i][j]>=1 && board[i][j]<=5) {
        			cctv.add(new int[] {board[i][j],i,j});
        		}
        	}
        }
        
        dfs(0,board);
        System.out.println(min);
    }

    static void search(int[] directions,int[][] board, int nx, int ny) {
    	for (int d : directions) {
    		int x=nx;
    		int y=ny;
			while(true) {
				if(x+dx[d]<0 || x+dx[d]>=N || y+dy[d]<0 || y+dy[d]>=M || board[x+dx[d]][y+dy[d]]==6  ) {
					break;
				}
				x+=dx[d];
				y+=dy[d];
				if(board[x][y]==0) {
					board[x][y]=7;
				}
			}
		}
    }
    
    static void dfs(int depth, int[][] board) {
    	if(depth==cctv.size()) {
    		int cnt=0;
    		for(int i=0; i<N; i++) {

    			for(int j=0; j<M; j++) {
    				if(board[i][j]==0) {
    					cnt++;
    				}
    			}
    		}
    		min=Math.min(min, cnt);

    		return;
    	}
    	
    	int[][] newBoard=new int[N][M];
    	for(int i=0;i<N;i++) {
    		for(int j=0;j<M;j++) {
    			newBoard[i][j]=board[i][j];
    		}
    	}
    	int t=cctv.get(depth)[0], x=cctv.get(depth)[1], y=cctv.get(depth)[2];
    	
    	
    	for(int[] d:dir[t]) {

    		search(d,newBoard,x,y);
    		dfs(depth+1,newBoard);

        	for(int i=0;i<N;i++) {
        		for(int j=0;j<M;j++) {
        			newBoard[i][j]=board[i][j];
        		}
        	}
    	}
    }

}